src/utilities/substituteFragmentArguments.ts

```diff
@@ -0,0 +1,86 @@
+import type { Maybe } from '../jsutils/Maybe.js';
+
+import type {
+  ArgumentNode,
+  FragmentDefinitionNode,
+  FragmentSpreadNode,
+  SelectionSetNode,
+  ValueNode,
+  VariableDefinitionNode,
+} from '../language/ast.js';
+import { Kind } from '../language/kinds.js';
+import { visit } from '../language/visitor.js';
+
+/**
+ * Replaces all fragment argument values with non-fragment-scoped values.
+ *
+ * NOTE: fragment arguments are scoped to the fragment they're defined on.
+ * Therefore, after we apply the passed-in arguments, all remaining variables
+ * must be either operation defined variables or explicitly unset.
+ */
+export function substituteFragmentArguments(
+  def: FragmentDefinitionNode,
+  fragmentSpread: FragmentSpreadNode,
+): SelectionSetNode {
+  const fragmentVariableDefinitions = def.variableDefinitions;
+  if (
+    fragmentVariableDefinitions == null ||
+    fragmentVariableDefinitions.length === 0
+  ) {
+    return def.selectionSet;
+  }
+
+  // We check whether we have arguments given by the fragment-spread
+  // we then replace our variable definitions in the fragment-definition
+  // with the values passed in by the spread.
+  const argumentValues = fragmentArgumentSubstitutions(
+    fragmentVariableDefinitions,
+    fragmentSpread.arguments,
+  );
+
+  return visit(def.selectionSet, {
+    Variable(node) {
+      return argumentValues.get(node.name.value);
+    },
+  });
+}
+
+export function fragmentArgumentSubstitutions(
+  variableDefinitions: ReadonlyArray<VariableDefinitionNode>,
+  argumentValues: Maybe<ReadonlyArray<ArgumentNode>>,
+): Map<string, ValueNode> {
+  const substitutions = new Map<string, ValueNode>();
+  if (argumentValues) {
+    for (const argument of argumentValues) {
+      substitutions.set(argument.name.value, argument.value);
+    }
+  }
+
+  for (const variableDefinition of variableDefinitions) {
+    const variableName = variableDefinition.variable.name.value;
+    if (substitutions.has(variableName)) {
+      continue;
+    }
+
+    const defaultValue = variableDefinition.defaultValue;
+    if (defaultValue) {
+      substitutions.set(variableName, defaultValue);
+    } else {
+      // We need a way to allow unset arguments without accidentally
+      // replacing an unset fragment argument with an operation
+      // variable value. Fragment arguments must always have LOCAL scope.
+      //
+      // TODO: To remove this hack, we need to either:
```

Comment:
The issue here presents itself in that fragments can define arguments like

```
fragment x($id: ID!) on Y { todo(id: $id) { text } }
```

Which can be passed in through fragment spreads

```
...x(id: '2')
```

now what happens when we spread `x` and there is an `$id` present on the document itself, if we aren't using `__UNSET` it would leverage that one which we don't want. There might be a better solution in carrying `variables` around as we visit the `FragmentSpread` and the accompanying `FragmentDefinition`